GTScript-Parallel-model.md

```diff
@@ -103,3 +103,175 @@ for k in range(start, end):
         u[i, j, k] = 1
     parfor [i_start:i_end, j_start:j_end]:
         b[i, j, k] = u[i-2,j,k] + u[i+1,j,k] + u[i,j-1,k] + u[i,j-2,k]
+```
+
+## Conditionals
+
+GTScript supports 2 kinds of conditionals:
+
+- conditionals on scalar expressions or compile-time constants 
+- conditionals on field expressions (meaning a condition on the field value at the gridpoint of the current iteration or at a local offset of it)
+
+### Conditionals on scalar expressions or compile-time constants
+
+- Conditions in `if` statements are evaluated at run-time and are required to be of scalar type, e.g. scalar variables (not fields) or scalar expressions (no fields involved).
+- Each statement inside the if and else branches is executed as a `parfor` loop (same as statements outside of branches).
+- There is no restriction on the body of the statement.
+- For consistency sake with the following we use the condition inside the loop as a concept (the implementation detail to not do this should not matter here)
+
+Rationale:
+- Adding an `if True:` does not change the behavior at all.
+- The way the programmer thinks about execution is the same if within an if or not.
+- This makes function calls and scoped code behave the same.
+
+#### Example
+
+```python
+with computation():
+    with interval(...):
+        if my_config_var:
+            a = 1
+            b = 2
+        else:
+            a = 2
+            b = 1
+```
+
+translates to: 
+
+```python
+for k in range(start, end):
+    parfor ij:
+        if my_config_var:
+            a[i, j, k] = 1
+    parfor ij:
+        if my_config_var:
+            b[i, j, k] = 2
+    parfor ij:
+        if not my_config_var:
+            a[i, j, k] = 2
+    parfor ij:
+        if not my_config_var:
+            b[i, j, k] = 1
+```
+
+### Conditionals on field expressions
+
+- Conditions in `if` statements are evaluated at run-time and are of storage type, e.g. fields or temporary variables.
+- The condition is evaluated for all gridpoints
+- Each statement inside the if and else branches is executed as a `parfor` loop (same as statements outside of branches).
+- Inside the if and else blocks the same field cannot be written to **and** read with an offset (order does not matter).
```

Comment:
Also, we should probably dissallow writing to any field implicated in the if-condition for the time being.

Reply:
> Also, we should probably dissallow writing to any field implicated in the if-condition for the time being.

On first read I thought the same, but it is not an issue as the condition is fully evaluated before the bodies as can be seen in the examples.

Reply:
> But then the question is whether the pseudo-code is actually what we implement or how we should think about it.

Both, that's actually the important point! Then, if we can we optimize.